dl/src/main/scala/com/intel/analytics/bigdl/dataset/Types.scala

```diff
@@ -152,7 +166,53 @@ trait Label[T] {
  * @param labels
  * @tparam T
  */
-case class MiniBatch[T](data: Tensor[T], labels: Tensor[T])
+case class MiniBatch[T: ClassTag](data: Activity, labels: Activity)
+  (implicit ev: TensorNumeric[T]) {
+
+  require(dataLength == labelLength,
+    "data and label batch size not match")
+
+  def this (pair: (Activity, Activity))
+  (implicit ev: TensorNumeric[T]) {
+    this(pair._1, pair._2)
+  }
+
+  def unapply[@specialized(Float, Double) T: ClassTag]
+  (batch: MiniBatch[T])
+  (implicit ev: TensorNumeric[T]): Option[(Activity, Activity)] =
+    Some((batch.data, batch.labels))
+
+  private def dataLength(): Int = {
+    data match {
+      case tensor: Tensor[T] => tensor.size(1)
+      case table: Table => table.length
+      case _ => throw new IllegalArgumentException("MiniBatch only supports Table or Tensor")
+    }
+  }
+
+  private def labelLength(): Int = {
+    labels match {
+      case tensor: Tensor[T] => tensor.size(1)
+      case table: Table => table.length
+      case _ => throw new IllegalArgumentException("MiniBatch only supports Table or Tensor")
+    }
+  }
+
+  def size(): Int = {
+    dataLength()
```

Comment:
should we get the dataLength when we construct it insteadOf calculate it every time?